src/clj/fluree/db/json_ld/transact.cljc

```diff
@@ -127,31 +136,30 @@
       (loop [[s-flakes & r] (partition-by flake/s new-flakes)
              subj-mods      {}]
         (if s-flakes
-          (let [sid            (flake/s (first s-flakes))
-                new-subject?   (first (filterv #(= const/$xsd:anyURI (flake/p %)) s-flakes))
-                new-classes    (->> s-flakes
-                                    (filterv #(= const/$rdf:type (flake/p %)))
-                                    (mapv flake/o))
-                classes        (if new-subject?
-                                 new-classes
-                                 (into new-classes
-                                       (<? (query-range/index-range db-before :spot = [sid const/$rdf:type]
-                                                                    {:flake-xf (map flake/o)}))))
-                class-shapes   (<? (shacl/class-shapes db-before classes))
+          (let [new-classes      (extract-classes s-flakes)
+                sid              (flake/s (first s-flakes))
+                existing-classes (<? (query-range/index-range db-before :spot = [sid const/$rdf:type]
+                                                              {:flake-xf (map flake/o)}))
+                classes          (into new-classes existing-classes)
+                class-shapes     (<? (shacl/class-shapes db-before classes))
                 ;; these target objects in s-flakes
-                pid->ref-flakes (when has-target-objects-of-shapes
-                                  (->> s-flakes
-                                       (filterv #(= const/$xsd:anyURI (flake/dt %)))
-                                       (group-by flake/p)))
-                o-pred-shapes   (when (seq pid->ref-flakes)
-                                  (<? (shacl/targetobject-shapes db-before (keys pid->ref-flakes))))
+                pid->ref-flakes  (when has-target-objects-of-shapes
+                                   (->> s-flakes
+                                        (filter (fn [f]
```

Comment:
I think the eager `filterv` is faster than `filter`, especially since we don't need the laziness here.